### PR TITLE
Feature/sprt 1871 use paginated table

### PR DESCRIPTION
**Pull request by dovekie**
@alexanderyoung PTAL

### CODE REVIEW

File: crittercism/models.py [hunk 1]
```diff
@@ -90,7 +91,22 @@ def latest_occurrences_by_version(self):
         }
 
     def build_date_map(self):
```
Comment (alexanderyoung) [inline crittercism/models.py]: Add a docstring please

Comment (alexanderyoung) [inline crittercism/models.py]: What about a unittest?

File: lib/etl/crashes.py [hunk 2]
```diff
@@ -9,8 +9,34 @@ def __init__(self, app_id, lookback_minutes, new_relic_account, new_relic_app_id
         super(CrashETL, self).__init__(app_id, lookback_minutes, new_relic_account, new_relic_app_id, self.EVENT_TYPE)
 
     def get_errors_with_details(self, lookback_minutes):
```
Comment (alexanderyoung) [inline lib/etl/crashes.py]: Docstring and unit test coverage?

File: lib/etl/crashes.py [hunk 3]
```diff
@@ -9,8 +9,34 @@ def __init__(self, app_id, lookback_minutes, new_relic_account, new_relic_app_id
...
+        daily_occurrences = {}
+        for version in versions:
+            versioned_summary = self._c_client.crash_paginated_tables(
+                self._app_id,
+                app_version=version,
+                lookback_timedelta=timedelta(minutes=lookback_minutes)
+            )
+            for error in versioned_summary[ERRORS]:
+                if daily_occurrences.get(error['hash']):
+                    daily_occurrences[error['hash']][version] = error['daily_occurrences']
```
Comment (alexanderyoung) [inline lib/etl/crashes.py]: Will we ever get here?

File: lib/etl/transactions.py [hunk 4]
```diff
@@ -4,29 +4,37 @@
...
+            'apteligent_userflow')
 
         period = 'PT%sM' % lookback_minutes
 
-        transactions_groups = self._c_client.transactions_details(self._app_id, period)
+        transactions_groups = self._c_client.transactions_details(self._app_id,
+                                                                  period)
 
-        groups = [[(g[u'name'], g[u'link']) for g in t[u'groups']] for t in transactions_groups]
+        groups = [[(g[u'name'], g[u'link'].replace('txn-report', 'developers'))
```
Comment (alexanderyoung) [inline lib/etl/transactions.py]: This is too complicated to read.  Please split it into a loop.

File: crittercism/client.py [hunk 5]
```diff
@@ -1,5 +1,5 @@
 import base64
-from datetime import datetime
+from datetime import datetime, timedelta
```
Comment (dovekie) [inline crittercism/client.py]: fixed

File: crittercism/client.py [hunk 6]
```diff
@@ -84,6 +86,7 @@ def __request_helper(self, verb, url, body_data, token, extra_headers=None):
 
         if extra_headers:
             headers.update(extra_headers)
+        url = url.replace(' ', '%20')
```
Comment (dovekie) [inline crittercism/client.py]: Fixed

File: crittercism/client.py [hunk 7]
```diff
@@ -195,6 +198,15 @@ def apps(self, attributes=None):
         response = [App(app_id, app_data) for app_id, app_data in content.items()]
         return response
 
+    def app_versions(self, app_id):
```
Comment (dovekie) [inline crittercism/client.py]: fixed

Comment (dovekie) [inline crittercism/client.py]: fixed

File: crittercism/client.py [hunk 8]
```diff
@@ -195,6 +198,15 @@ def apps(self, attributes=None):
...
+        url_suffix = 'apps?attributes=appVersions'
```
Comment (dovekie) [inline crittercism/client.py]: fixed

File: crittercism/client.py [hunk 9]
```diff
@@ -195,6 +198,15 @@ def apps(self, attributes=None):
...
+        content = self.__request('GET', url_suffix, {})
```
Comment (dovekie) [inline crittercism/client.py]: What's the most efficient way to use these inside a set of classes like these?

File: crittercism/client.py [hunk 10]
```diff
@@ -227,6 +239,24 @@ def app_crash_counts(self, app_id):
 
         return content
 
+    def crash_paginated_tables(self, app_id, app_version=None,
+                               lookback_timedelta=None):
+        url_suffix = '{}/crash/paginatedtable'.format(app_id)
+
+        if app_version:
+            url_suffix += u'?appVersion={}&'.format(app_version)
```
Comment (dovekie) [inline crittercism/client.py]: Fixed

File: crittercism/client.py [hunk 11]
```diff
@@ -227,6 +239,24 @@ def app_crash_counts(self, app_id):
...
+
+        if lookback_timedelta:
+            start_date = (datetime.now() - lookback_timedelta).isoformat()
+            end_date = datetime.now().isoformat()
+
+            url_suffix += 'startDate={}&endDate={}'.format(start_date,
+                                                            end_date)
+
+        content = self.__request('GET', url_suffix, {})
```
Comment (dovekie) [inline crittercism/client.py]: see above regarding string literals in classes

File: lib/etl/transactions.py (hunk 4 above)
Comment (dovekie) [inline lib/etl/transactions.py]: Fixed

File: crittercism/client.py [hunk 12]
```diff
@@ -227,6 +239,24 @@ def app_crash_counts(self, app_id):
 
         return content
 
+    def crash_paginated_tables(self, app_id, app_version=None,
+                               lookback_timedelta=None):
```
Comment (dovekie) [inline crittercism/client.py]: fixed

Comment (dovekie) [inline crittercism/client.py]: fixed

File: crittercism/models.py [hunk 13]
```diff
@@ -82,16 +90,47 @@ def date_occurrences(self, date):
...
+                        date = (start_date - timedelta(days=index))
+                        total[date] += error_count
+
+                dates = sorted(total.keys())
+                total_daily_occurrences = (dates[-1].strftime('%Y-%m-%d'), [])
+
+                for date in dates:
+                    total_daily_occurrences[1].append(total[date])
+            else:
+                return None
```
Comment (dovekie) [inline crittercism/models.py]: Yes, it should return a datetime object and a dictionary of datetimes and error totals.

File: crittercism/models.py [hunk 14]
```diff
@@ -82,16 +90,47 @@ def date_occurrences(self, date):
...
             version: num_list[0] if date == todays_date else 0
             for version, [date, num_list]
-            in self._data[u'dailyOccurrencesByVersion'].items()
+            in self._data['daily_occurrences_by_version'].items()
         }
 
     def build_date_map(self):
-        incomplete_data_latest_date, daily_data = self._data[u'dailyOccurrencesByVersion'][u'total']
-        latest_complete_datetime = datetime.strptime(incomplete_data_latest_date, '%Y-%m-%d')
+        """
```
Comment (dovekie) [inline crittercism/models.py]: fixed

File: crittercism/models.py [hunk 15]
```diff
@@ -82,16 +90,47 @@ def date_occurrences(self, date):
...
+        For crashes, this uses the array of arrays in 'daily_occourrences'
```
Comment (dovekie) [inline crittercism/models.py]: fixed

File: crittercism/models.py [hunk 16]
```diff
@@ -107,36 +146,50 @@ def build_date_map(self):
         return latest_complete_datetime, date_map
 
     def unique_session_count(self):
-        return self._data[u'uniqueSessionCountsByVersion'][u'total']
+        return self._data['num_unique_sessions']
```
Comment (dovekie) [inline crittercism/models.py]: fixed

File: crittercism/models.py [hunk 17]
```diff
@@ -107,36 +146,50 @@ def build_date_map(self):
...
-            'unsymbolizedHash': self._data[u'unsymbolizedHash'],
-            'firstOccurred': self._data[u'firstOccurred'],
-            'appType': self._data[u'appType'],
-            'name': self._data[u'name'],
-            'reason': self._data[u'reason'],
-            'displayReason': self._data[u'displayReason'],
-        }
+        return self._data['appId']
+
+    def as_event_dict(self, app_id):
```
Comment (dovekie) [inline crittercism/models.py]: fixed

File: crittercism/models.py [hunk 18]
```diff
@@ -107,36 +146,50 @@ def build_date_map(self):
...
+                'firstOccurred': self._data['first_occurred_time'],
+                'name': self._data['name'],
+                'reason': self._data['reason'],
+                'displayReason': self._data['display_reason'],
+                }
+        else:
+            return {
+                'crAppId': self._data[u'appId'],
+                'hash': self._data[u'hash'],
+                'unsymbolizedHash': self._data[u'unsymbolizedHash'],
```
Comment (dovekie) [inline crittercism/models.py]: fixed

File: lib/etl/crashes.py (hunk 2 above)
Comment (dovekie) [inline lib/etl/crashes.py]: fixed

File: crittercism/models.py (hunk 1 above)
Comment (dovekie) [inline crittercism/models.py]: Fixed

File: crittercism/client.py (hunk 9 above)
Comment (dovekie) [inline crittercism/client.py]: I believe this is fixed

Comment (alexanderyoung) [inline crittercism/client.py]: As discussed, please add them at the top of the class def

File: crittercism/client.py [hunk 19]
```diff
@@ -195,6 +198,15 @@ def apps(self, attributes=None):
         response = [App(app_id, app_data) for app_id, app_data in content.items()]
         return response
 
+    def app_versions(self, app_id):
+        url_suffix = 'apps?attributes=appVersions'
+        content = self.__request('GET', url_suffix, {})
+
+        versions = content[app_id]['appVersions']
```
Comment (alexanderyoung) [inline crittercism/client.py]: Need an answer

File: crittercism/client.py (hunk 12 above)
Comment (alexanderyoung) [inline crittercism/client.py]: Still no docstring here?

File: crittercism/client.py (hunk 10 above)
Comment (alexanderyoung) [inline crittercism/client.py]: You say this is fixed, but I don't see the change.

File: crittercism/client.py (hunk 19 above)
Comment (dovekie) [inline crittercism/client.py]: fixed

File: crittercism/client.py (hunk 12 above)
Comment (dovekie) [inline crittercism/client.py]: fixed

Comment (dovekie) [inline crittercism/client.py]: fixed

File: crittercism/models.py [hunk 20]
```diff
@@ -54,10 +55,17 @@ class CRErrorBase(object):
     def __init__(self, hash_init):
         self._data = hash_init
         self._oldest_available = None
-        self._tz_offset = self._data[u'appTimeZoneOffset']
+        if self.is_crash:
+            self._tz_offset = self._data.get('app_timezone_offset')
+        else:
+            self._tz_offset = self._data.get('appTimeZoneOffset')
```
Comment (dovekie) [inline crittercism/models.py]: fixed

File: crittercism/models.py (hunk 14 above)
Comment (dovekie) [inline crittercism/models.py]: fixed

File: crittercism/models.py (hunk 13 above)
Comment (dovekie) [inline crittercism/models.py]: Yes, this should return a datetime of the last complete date and a dict of total occurrences on each day

File: crittercism/models.py (hunk 16 above)
Comment (dovekie) [inline crittercism/models.py]: fixed

File: crittercism/models.py (hunk 17 above)
Comment (dovekie) [inline crittercism/models.py]: fixed

Comment (dovekie) [inline crittercism/models.py]: fixed

File: crittercism/models.py [hunk 21]
```diff
@@ -107,36 +146,50 @@ def build_date_map(self):
...
-            'appType': self._data[u'appType'],
-            'name': self._data[u'name'],
-            'reason': self._data[u'reason'],
-            'displayReason': self._data[u'displayReason'],
-        }
+        return self._data['appId']
+
+    def as_event_dict(self, app_id):
+        if self.is_crash:
+            return {
```
Comment (dovekie) [inline crittercism/models.py]: fixed

File: crittercism/models.py (hunk 18 above)
Comment (dovekie) [inline crittercism/models.py]: fixed

File: lib/etl/crashes.py (hunk 2 above)
Comment (dovekie) [inline lib/etl/crashes.py]: fixed

File: tests/helpers.py [hunk 22]
```diff
@@ -48,4 +49,16 @@ def __init__(self, *args, **kw):
 
         self.add_response_ignore_body_mapping('POST',
                                               'https://insights-collector.newrelic.com/v1/accounts/1/events',
-                                              200, {})
\ No newline at end of file
+                                              200, {})
+#
```
Comment (alexanderyoung) [inline tests/helpers.py]: Do you need this comment?

File: tests/helpers.py [hunk 23]
```diff
@@ -11,7 +11,8 @@ def _response(response_code):
 
 
 def _response_map_key(verb, url, body=None):
-    return '%s|%s|%s' % (verb, url, body)
+    print "response map", '{}|{}|{}'.format(verb, url, body)
```
Comment (alexanderyoung) [inline tests/helpers.py]: Is this print line for testing?

File: tests/helpers.py [hunk 24]
```diff
@@ -31,7 +32,7 @@ def add_response_mapping(self, verb, url, response_code, response_content, reque
                                               json.dumps(request_body or {}))] = (response_code, response_content)
 
     def request(self, url, verb, headers=None, body=None):
-        logging.getLogger().debug("MockHTTP requesting url=%s verb=%s headers=%s body=%s", url, verb, headers, body)
+        logging.getLogger().warn("MockHTTP requesting url=%s verb=%s headers=%s body=%s", url, verb, headers, body)
```
Comment (alexanderyoung) [inline tests/helpers.py]: Why is this a warning?

File: tests/helpers.py (hunk 23 above)
Comment (dovekie) [inline tests/helpers.py]: it was probably for debugging. Removed.

File: tests/helpers.py (hunk 24 above)
Comment (dovekie) [inline tests/helpers.py]: To make it stand out while I was working on it. Fixed.

File: crittercism/models.py [hunk 25]
```diff
@@ -82,16 +107,47 @@ def date_occurrences(self, date):
...
+        For crashes, this uses the list of lists in 'daily_occurrences'
+        For exceptions, adds the 'dailyOccurrencesByVersion' for each version
+        to get the total daily occurrences.
+
+        :return: The datetime of the last day of complete data,
+        and a dictionary of datetimes with the total number of occurrences
+        of this crash on that day.
+        """
+        if self.is_crash:
+            incomplete_data_latest_date, daily_data = self._data['daily_occurrences']
```
Comment (alexanderyoung) [inline crittercism/models.py]: String literal

File: crittercism/models.py [hunk 26]
```diff
@@ -82,16 +107,47 @@ def date_occurrences(self, date):
...
+            latest_complete_datetime = datetime.strptime(
+                incomplete_data_latest_date, '%Y%m%d')
+        else:
+            daily_occurrences_by_version = self._data.get('dailyOccurrencesByVersion')
```
Comment (alexanderyoung) [inline crittercism/models.py]: String literal

File: crittercism/models.py [hunk 27]
```diff
@@ -82,16 +107,47 @@ def date_occurrences(self, date):
...
+            if daily_occurrences_by_version:
+                incomplete_data_latest_date, daily_data = daily_occurrences_by_version['total']
```
Comment (alexanderyoung) [inline crittercism/models.py]: String literal

File: crittercism/models.py [hunk 28]
```diff
@@ -82,16 +107,47 @@ def date_occurrences(self, date):
...
+                latest_complete_datetime = datetime.strptime(incomplete_data_latest_date, '%Y-%m-%d')
```
Comment (alexanderyoung) [inline crittercism/models.py]: String literal

File: crittercism/models.py [hunk 29]
```diff
@@ -82,16 +107,47 @@ def date_occurrences(self, date):
...
+                total = defaultdict(int)
+                for version, errors in daily_occurrences_by_version.items():
+                    start_date = datetime.strptime(errors[0], '%Y-%m-%d')
```
Comment (alexanderyoung) [inline crittercism/models.py]: String literal

File: crittercism/models.py [hunk 30]
```diff
@@ -82,16 +107,47 @@ def date_occurrences(self, date):
...
+                    daily_error_counts = errors[1]
+                    for index, error_count in enumerate(reversed(daily_error_counts)):
+                        date = (start_date - timedelta(days=index))
+                        total[date] += error_count
+
+                dates = sorted(total.keys())
+                total_daily_occurrences = (dates[-1].strftime('%Y-%m-%d'), [])
```
Comment (alexanderyoung) [inline crittercism/models.py]: String literal

File: crittercism/models.py [hunk 31]
```diff
@@ -107,36 +163,55 @@ def build_date_map(self):
...
+        Convert the error object into a dictionary
+
+        :param app_id: string
+        :return: dict
+        """
+        if self.is_crash:
+            return {
+                self.cr_app_id: app_id,
+                self._hash: self._data[self._hash],
+                self.first_occurred: self._data['first_occurred_time'],
```
Comment (alexanderyoung) [inline crittercism/models.py]: String literal

File: crittercism/models.py [hunk 32]
```diff
@@ -107,36 +163,55 @@ def build_date_map(self):
...
+                self.name: self._data[self.name],
+                self.reason: self._data[self.reason],
+                self.display_reason: self._data['display_reason'],
```
Comment (alexanderyoung) [inline crittercism/models.py]: String literal

File: crittercism/models.py [hunk 33]
```diff
@@ -178,15 +260,21 @@ def add_optional_params(self, params):
...
-        self._start_time = datetime.strptime(self._response[u'start'], '%Y-%m-%dT%H:%M:%S')
-        self._end_time = datetime.strptime(self._response[u'end'], '%Y-%m-%dT%H:%M:%S')
+        self._response = response[self.data]
+        self._slices = self._response.get('slices', [])
+        self._total_pie = sum([s[self.value] for s in self._slices])
+        self._slices_by_name = {s[self.name]: s[self.value] for s in self._slices}
+        self._slices_by_label = {s[self.label]: s[self.value] for s in self._slices}
+        self._group_list = [(s[self.name], s[self.label]) for s in self._slices]
+        self._start_time = datetime.strptime(self._response['start'],
+                                             '%Y-%m-%dT%H:%M:%S')
```
Comment (alexanderyoung) [inline crittercism/models.py]: String literal

File: crittercism/models.py [hunk 34]
```diff
@@ -178,15 +260,21 @@ def add_optional_params(self, params):
...
+        self._end_time = datetime.strptime(self._response['end'],
+                                           '%Y-%m-%dT%H:%M:%S')
```
Comment (alexanderyoung) [inline crittercism/models.py]: String literal

File: lib/etl/crashes.py [hunk 35]
```diff
@@ -9,8 +9,49 @@ def __init__(self, app_id, lookback_minutes, new_relic_account, new_relic_app_id
...
+            )
+            for error in versioned_summary[DATA][ERRORS]:
+                if daily_occurrences.get(error[HASH]):
+                    daily_occurrences[error[HASH]][version] = error[DAILY_OCCURRENCES]
+
+                else:
+                    daily_occurrences[error[HASH]] = {version: error[DAILY_OCCURRENCES]}
+        crashes = []
+        for summary in crash_summaries[DATA][ERRORS]:
+            summary['daily_occurrences_by_version'] = daily_occurrences.get(
```
Comment (alexanderyoung) [inline lib/etl/crashes.py]: String literal

File: lib/etl/transactions.py [hunk 36]
```diff
@@ -4,29 +4,49 @@
...
+                traces_urls.append((group_name, url))
+
+        for group_name, url in traces_urls:
+            pages = self._c_client.get_paged_transaction_data(app_id, url)
 
-                for page in self._c_client.get_paged_transaction_data(app_id, url):
-                    for trace in page[u'traces']:
-                        self._events.append(self.make_event(trace, group_name))
+            for page in pages:
+                for trace in page['data']:
```
Comment (alexanderyoung) [inline lib/etl/transactions.py]: String literal

File: lib/etl/transactions.py [hunk 37]
```diff
@@ -4,29 +4,49 @@
...
+                    self._events.append(self.make_event(trace, group_name))
 
     def make_event(self, trace, group_name):
         dt = datetime.strptime(trace[u'traceTs'], '%Y-%m-%dT%H:%M:%S.%fZ')
```
Comment (alexanderyoung) [inline lib/etl/transactions.py]: String literal

File: lib/new_relic/client.py [hunk 38]
```diff
@@ -32,30 +33,49 @@ def __request(self, verb, body_data):
...
-        response, content = self._http.request(self.NEW_RELIC_URL % (self.NEW_RELIC_API_DOMAIN, self.account_id),
-                                               verb,
-                                               headers=headers,
-                                               body=json.dumps(body_data))
+        logging.getLogger().debug('Making request to: {}'.format(url))
 
-        return json.loads(content)
+        response = None
+
+        if verb == 'GET':
```
Comment (alexanderyoung) [inline lib/new_relic/client.py]: String literal

File: lib/new_relic/client.py [hunk 39]
```diff
@@ -32,30 +33,49 @@ def __request(self, verb, body_data):
...
+            response = requests.get(url, headers=headers)
+        elif verb == 'POST':
```
Comment (alexanderyoung) [inline lib/new_relic/client.py]: String literal

**Pull request merged**